File: workspaces-good/modules/cli/src/surface_calculator_command.py
import sys
from typing import List

from shared.src.contract_calculator_aggregate import CalculatorAggregate
from shared.src.taxonomy_expression_vo import create_expression
from shared.src.taxonomy_operation_vo import operation_from_symbol


def run(calc: CalculatorAggregate) -> None:
    print("=== Calculator ===", file=sys.stderr)
    print("Ketik operasi: <angka> <operator> <angka>", file=sys.stderr)
    print("Contoh: 2 + 3", file=sys.stderr)
    print("Ketik 'h' untuk riwayat, 'q' untuk keluar", file=sys.stderr)

    while True:
        try:
            line = input("> ")
        except EOFError:
            break
        trimmed = line.strip()
        if trimmed == "q":
            break
        if trimmed == "h":
            hist: List = calc.history()
            if not hist:
                print("  Belum ada riwayat", file=sys.stderr)
            else:
                for r in hist:
                    print(f"  {r.expression}", file=sys.stderr)
            continue
        parts = trimmed.split()
        if len(parts) != 3:
            print("  Format: <angka> <operator> <angka>", file=sys.stderr)
            continue
        try:
            left = float(parts[0])
            right = float(parts[2])
        except ValueError:
            print("  Input bukan angka", file=sys.stderr)
            continue
        op = operation_from_symbol(parts[1])
        if op is None:
            print("  Operator tidak dikenal", file=sys.stderr)
            continue
        expr = create_expression(left, op, right)
        result = calc.delegate(expr)
        if result:
            print(f"  = {result.value}", file=sys.stderr)
        else:
            print("  Error: tidak bisa hitung", file=sys.stderr)
    print("Sampai jumpa!", file=sys.stderr)

File: workspaces-good/packages/cli/src/surface_calculator_command.ts
import { CalculatorAggregate } from "calculator-shared/src/contract_calculator_aggregate";
import { ExpressionVO, createExpression } from "calculator-shared/src/taxonomy_expression_vo";
import { OperationVO, operationFromSymbol } from "calculator-shared/src/taxonomy_operation_vo";
import * as readline from "readline";

export function run(calc: CalculatorAggregate): void {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stderr,
  });
  console.error("=== Calculator ===");
  console.error("Ketik operasi: <angka> <operator> <angka>");
  console.error("Contoh: 2 + 3");
  console.error("Ketik 'h' untuk riwayat, 'q' untuk keluar");

  const prompt = () => {
    rl.question("> ", (input) => {
      const trimmed = input.trim();
      if (trimmed === "q") {
        console.error("Sampai jumpa!");
        rl.close();
        return;
      }
      if (trimmed === "h") {
        const hist = calc.history();
        if (hist.length === 0) {
          console.error("  Belum ada riwayat");
        } else {
          hist.forEach((r) => console.error(`  ${r.expression}`));
        }
        prompt();
        return;
      }
      const parts = trimmed.split(/\s+/);
      if (parts.length !== 3) {
        console.error("  Format: <angka> <operator> <angka>");
        prompt();
        return;
      }
      const left = parseFloat(parts[0]);
      const op = operationFromSymbol(parts[1]);
      const right = parseFloat(parts[2]);
      if (isNaN(left) || isNaN(right)) {
        console.error("  Input bukan angka");
        prompt();
        return;
      }
      if (op === null) {
        console.error(`  '${parts[1]}' bukan operator valid`);
        prompt();
        return;
      }
      const expr: ExpressionVO = createExpression(left, op, right);
      const result = calc.delegate(expr);
      if (result) {
        console.error(`  = ${result.value}`);
      } else {
        console.error("  Error: tidak bisa hitung");
      }
      prompt();
    });
  };
  prompt();
}
